[Jerome Pui] iP

jeromepui opened the pull request:
### Jarvis

> “Life is too complicated not to be orderly.” – Martha Stewart ([source](https://dansilvestre.com/productivity-quotes/#10_Life_is_too_complicated_not_to_be_orderly_-_Martha_Stewart))

Jarvis, your reliable personal assistant, keeps you on track by managing your tasks, deadlines and events. It's

- text-based
- intuitive
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is:

1. download it from [here](https://github.com/jeromepui/ip/releases/tag/A-Jar).
2. launch the application
3. manage your tasks, deadlines and events 😉

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events

And most importantly, it is **FREE**!

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Jarvis jarvis = new Jarvis();
        jarvis.run();
    }
}
```

Commit:
Add Gradle support

jeromepui committed:
Add greet, echo and exit features

jeromepui committed:
Add ability to create tasks and list them

jeromepui committed:
Add Task class

jeromepui committed:
Add ability to mark tasks as done or not done

jeromepui committed:
Rename Duke to Jarvis

jeromepui committed:
Add Command class

jeromepui committed:
Add ByeCommand, ListCommand and MarkCommand classes

jeromepui committed:
Add support for tracking ToDos, Events and Deadlines

jeromepui committed:
Semi-automate testing of Jarvis

jeromepui committed:
Add JarvisException class

jeromepui committed:
Handle basic errors

jeromepui committed:
Add support for deleting tasks

jeromepui committed:
Save tasks in hard disk automatically

jeromepui committed:
Format dates and times using LocalDate and LocalDateTime

jeromepui committed:
Merge branch 'branch-Level-7'

jeromepui committed:
Merge branch 'branch-Level-8'

jeromepui committed:
Add Ui, Parser, TaskList classes

jeromepui committed:
Divide classes into packages

jeromepui committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

jeromepui committed:
Add JUnit tests

jeromepui committed:
Add JavaDoc comments

jeromepui committed:
Tweak code to comply with coding standard

jeromepui committed:
Add FindCommand class

jeromepui committed:
Merge branch 'branch-A-JavaDoc'

jeromepui committed:
Merge branch 'branch-A-CodingStandard'

jeromepui committed:
Merge branch 'branch-Level-9'

cadencjk commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,56 @@
+package jarvis;
+
+import java.util.Scanner;
```
Great job for week 3! Most of the code adheres to coding standards and OOP, but comments could be elaborated more. Keep up the good work 😄

cadencjk commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,56 @@
+        this.ui.greet();
+
+        Scanner sc = new Scanner(System.in);
+        boolean terminated = false;
```
Got your idea that the Boolean tracks whether the chat is still ongoing.
Should the Boolean name be a verb instead?
```suggestion
        boolean isTerminated = false;
````

cadencjk commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,56 @@
+        while (!terminated) {
+            try {
+                Command command = Parser.parseUserCommand(sc);
+                String keyCommand = command.getKeyCommand();
```
Could Enums be used to represent the type of keyCommand?

cadencjk commented on src/main/java/jarvis/command/Command.java:
```diff
@@ -0,0 +1,78 @@
+
+    String[] splitArguments() {
+        return arguments.map(s -> Arrays.stream(s.split("/by|/at", 2)).map(String::trim)
+                .toArray(String[]::new)).orElse(null);
```
Nice use of Java pipelines to generate functions 👍!

cadencjk commented on src/main/java/jarvis/command/FindCommand.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws JarvisException {
+        List<Task> found = tasks.findTasks(super.getDescription());
```
Perhaps you could rename your List to specify the type of items found?
```suggestion
        List<Task> tasksFound = tasks.findTasks(super.getDescription());
````

cadencjk commented on src/main/java/jarvis/task/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+ * Deadline --- deadline task.
+ */
+public class Deadline extends Task {
+    private final LocalDate by;
```
Should attributes be named as verbs? 👀

cadencjk reviewed (commented):
LGTM 👍
More detailed comments could be provided.

jeromepui committed:
Implement GUI using JavaFX

jeromepui committed:
Use assertions in Command and Parser classes

Use assert feature to document important assumptions that should hold at
various points in the code to reduce potential errors that may occur in
the future.

jeromepui committed:
Improve code quality

Some of the code does not follow coding standard guidelines. For
example, some methods and variables were named vaguely so they have been
renamed such that they are less confusing for the reader.

jeromepui committed:
Improve code quality

Some of the code does not follow coding standard guidelines. For
example, some methods and variables were named vaguely so they have been
renamed such that they are less confusing for the reader.

jeromepui committed:
Merge pull request #1 from jeromepui/branch-A-Assertions

Use assertions in Command and Parser classes

jeromepui committed:
Merge branch 'master' into branch-A-CodeQuality

jeromepui committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/jeromepui/ip into branch-A-CodeQuality

jeromepui committed:
Merge pull request #2 from jeromepui/branch-A-CodeQuality

Improve code quality

jeromepui committed:
Add ability to sort tasks by deadline and start times

jeromepui committed:
Merge branch 'branch-C-Sort'

jeromepui committed:
Improve the GUI

jeromepui committed:
Refactor code

jeromepui committed:
Add user guide

jeromepui committed:
Fix Gradle bug for release
